tests/unit/s2n_self_talk_psk_test.c

```diff
@@ -0,0 +1,340 @@
+
+        /* Clean-up */
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+        EXPECT_SUCCESS(s2n_config_free(config));
+        EXPECT_SUCCESS(s2n_io_pair_close(&io_pair));
+        EXPECT_SUCCESS(s2n_cert_chain_and_key_free(chain_and_key));
+    }
+
+    /* Happy Case: PSK with callback function */ 
+    {
+        /* Setup connections */
+        struct s2n_connection *client_conn = NULL;
+        struct s2n_connection *server_conn = NULL;
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+
+        /* Setup config */
+        struct s2n_cert_chain_and_key *chain_and_key = NULL;
+        EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+                S2N_DEFAULT_ECDSA_TEST_CERT_CHAIN, S2N_DEFAULT_ECDSA_TEST_PRIVATE_KEY));
+        struct s2n_config *config = NULL;
+        EXPECT_NOT_NULL(config = s2n_config_new());
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default_tls13"));
+        EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        /* Create nonblocking pipes */
+        struct s2n_test_io_pair io_pair = { 0 };
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
+
+        /* Setup PSKs */
+        EXPECT_OK(setup_client_psks(client_conn));
+        EXPECT_OK(setup_server_psks(server_conn));
+
+        /* Set the customer callback to select PSK identity */ 
+        EXPECT_SUCCESS(s2n_config_set_psk_selection_callback(server_conn, s2n_test_select_psk_identity_callback));
+        EXPECT_EQUAL(server_conn->config->psk_selection_cb, s2n_test_select_psk_identity_callback);
+
+        /* Negotiate handshake */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+        /* Validate handshake type */
+        EXPECT_FALSE(client_conn->handshake.handshake_type & CLIENT_AUTH);
+        EXPECT_FALSE(server_conn->handshake.handshake_type & CLIENT_AUTH);
+        EXPECT_FALSE(client_conn->handshake.handshake_type & FULL_HANDSHAKE);
+        EXPECT_FALSE(server_conn->handshake.handshake_type & FULL_HANDSHAKE);
+
+        /* Validate chosen PSK */
+        EXPECT_EQUAL(server_conn->psk_params.chosen_psk->identity.size, sizeof(test_shared_identity));
+        EXPECT_EQUAL(memcmp(server_conn->psk_params.chosen_psk->identity.data, test_shared_identity,
+                            sizeof(test_shared_identity)), 0);
```

Comment:
How do you know your callback was triggered? I'd test this with a callback written specifically for this test, not a clone of the default logic. Choose an identity the default logic wouldn't choose, maybe also set some value we can check.